aspnetcore/fundamentals/logging/sample/Controllers/TodoController.cs

```diff
@@ -0,0 +1,383 @@
+﻿#define LoggerDI // or CreateLogger or LogException or Scopes
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using TodoApi.Core;
+using TodoApi.Core.Interfaces;
+using TodoApi.Core.Model;
+using System;
+
+namespace TodoApi.Controllers
+{
+#if LoggerDI
+    [Route("api/[controller]")]
+    #region snippet_LoggerDI
+    public class TodoController : Controller
+    {
+        private readonly ITodoRepository _todoRepository;
+        private readonly ILogger _logger;
+
+        public TodoController(ITodoRepository todoRepository,
+            ILogger<TodoController> logger)
+        {
+            _todoRepository = todoRepository;
+            _logger = logger;
+        }
+        #endregion
+        [HttpGet]
+        public IEnumerable<TodoItem> GetAll()
+        {
+            using (_logger.BeginScope("Message {HoleValue}", DateTime.Now))
+            {
+                _logger.LogInformation(LoggingEvents.ListItems, "Listing all items");
+                EnsureItems();
+            }
+            return _todoRepository.GetAll();
+        }
+
+        [HttpGet("{id}", Name = "GetTodo")]
+        #region snippet_CallLogMethods
+        public IActionResult GetById(string id)
+        {
+            _logger.LogInformation(LoggingEvents.GetItem, "Getting item {ID}", id);
+            var item = _todoRepository.Find(id);
+            if (item == null)
+            {
+                _logger.LogWarning(LoggingEvents.GetItemNotFound, "GetById({ID}) NOT FOUND", id);
+                return NotFound();
+            }
+            return new ObjectResult(item);
+        }
+        #endregion
+
+        [HttpPost]
+        public IActionResult Create([FromBody] TodoItem item)
+        {
+            if (item == null)
+            {
+                return BadRequest();
+            }
+            _todoRepository.Add(item);
+            _logger.LogInformation(LoggingEvents.InsertItem, "Item {0} Created", item.Key);
```

Comment:
Leftover from original version of the article, I'll change to named parameter syntax.